qiskit/transpiler/passes/scheduling/padding/context_aware_dynamical_decoupling.py

```diff
@@ -0,0 +1,756 @@
+            """Add another delay event to an existing block to either extend or close it."""
+            open_delay_block.events.append(delay_event)
+
+            if delay_event.type == EventType.BEGIN:  # TODO: this is never called
+                open_delay_block.active_qubits += set(delay_event.op_node.qargs)
+            else:
+                open_delay_block.active_qubits -= set(delay_event.op_node.qargs)
+
+            if not open_delay_block.active_qubits:
+                open_delay_blocks.remove(open_delay_block)
+                closed_delay_blocks.append(open_delay_block)
+
+        def _combine_delay_blocks(delay_blocks):
+            survivor, *doomed = delay_blocks
+
+            for doomed_delay_group in doomed:
+                # Add events and qubits from doomed block to survivor.
+                if survivor.active_qubits.intersection(doomed_delay_group.active_qubits):
+                    raise RuntimeError("More than one open delay on a qubit?")
+
+                survivor.events.extend(doomed_delay_group.events)
+                survivor.active_qubits.update(doomed_delay_group.active_qubits)
+
+                open_delay_blocks.remove(doomed_delay_group)
+                survivor.events.sort(key=DelayEvent.sort_key)  # Maintain sorted event order
+
+        for delay_event in sorted_delay_events:
+            # This could be avoided by keeping a map of device qubit to open
+            # block and only considering neighbors of current event.
+            # use coupling_map.graph.neighbors_undirected once Qiskit/rustworkx#1254 is in a release
+            adjacent_open_delay_blocks = [
+                open_delay
+                for open_delay in open_delay_blocks
+                if any(
+                    self._coupling_map.distance(
+                        qubit_map[delay_event.op_node.qargs[0]], qubit_map[open_delay_qubit]
+                    )
+                    <= 1
+                    for open_delay_qubit in open_delay.active_qubits
+                )
+            ]
+
+            if delay_event.type == EventType.BEGIN:
+                # If crossing a begin edge, check if there are any open delays that are adjacent.
+                # If so, add current event to that group.
+
+                if len(adjacent_open_delay_blocks) == 0:
+                    # Make a new delay
+                    _open_delay_block(delay_event)
+                else:
+                    # Make a new block and combine that with adjacent open blocks
+                    new_block = _open_delay_block(delay_event)
+                    _combine_delay_blocks(adjacent_open_delay_blocks + [new_block])
+
+            if delay_event.type == EventType.END:
+                # If crossing a end edge, remove this qubit from the actively delaying qubits"
+                if len(adjacent_open_delay_blocks) != 1:
+                    raise RuntimeError("Closing edge w/o an open delay?")
```

Comment:
I agree with Eli here about the intent of these exception not being clear or user actionable, but I don't think a runtime assertion is a good choice. 

Using `assert` is not really desirable either because the raised `AssertionError` is also not really actionable and typically ends up being even more cryptic, especially since there is little control over the exception raised. I'd argue either change the exceptions raised to a qiskit exception type and change the error message to indicate how a user can address the failure mode or just remove the check completely. If it's just an internal sanity check for debugging we should remove it and avoid the runtime overhead completely. In theory this is what you can do with asserts, because `python -O` will strip the asserts but in practice I've very rarely seen anyone use that flag.

Reply:
Fwiw, we have used `RuntimeError` for other places in Qiskit to represent an internal logic error, but the idea is that it should be entirely impossible for a user to trigger them (unless they're doing something they're not allowed to), or they're a flag to implementers that they've done something wrong.

That said, I believe that the majority of places it's used in Qiskit already are places where it has no runtime cost to do, because it's the "default" case of a lookup table or an `if/else` chain, or something like that, and we need to do _something_ other than just fall off the end of the chain silently.  I think quite a few of the uses here don't fall into those buckets, and should be replaced with proper errors - some of them are user actionable, and some are internal logic validation that has non-trivial overhead to check.

Reply:
I opted for the third variant: adding optional `logger.debug` statements, if the respective logging level is enabled. This code has been tested on dummy circuits and a variety of smaller hardware experiments, but I'd prefer keeping an easy way of printing debug info as long as the code is relatively new. Let me know if this is fine, otherwise I can also remove these!

Reply:
Yes, a logging approach is good